segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java

```diff
@@ -0,0 +1,457 @@
+                           }
+                           return null;
+                       });
+        }
+
+        /**
+         * Gets a delayed future when either i) an error has been experienced in the last iteration, or ii) if there is
+         * no outstanding data to process. Otherwise, we do not delay the next iteration and return a completed future.
+         *
+         * @return A future that may delay the next processing iteration if there has been an error in the last
+         * processing iteration or there is no outstanding data to process.
+         */
+        private CompletableFuture<Void> getDelayedFutureIfNeeded() {
+            return failedIteration.get() || outstandingBytes.get() == 0 ? Futures.delayedFuture(iterationDelay, executor) :
+                    Futures.delayedFuture(Duration.ZERO, executor);
+        }
+
+        /**
+         * Returns a {@link CompletableFuture} that results from the execution of the following tasks for each EventProcessor:
+         * i) Read available data in the Segment, ii) deserialize at most {@link EventProcessorConfig#getMaxItemsAtOnce()},
+         * iii) execute the handler function for the {@link EventProcessor}, and iv) truncate the internal Segment.
+         */
+        private CompletableFuture<Void> processEvents() {
+            final Timer iterationTime = new Timer();
+            return readEvents()
+                    .thenComposeAsync(this::applyProcessorHandler, executor)
+                    .thenComposeAsync(readResult -> truncateInternalSegment(readResult, iterationTime), executor)
+                    .handleAsync((r, ex) -> {
+                        // If we got an exception different from NoDataAvailableException, report it as something is off.
+                        if (ex != null && !(Exceptions.unwrap(ex) instanceof NoDataAvailableException)) {
+                            log.warn("{}: Processing iteration failed, retrying.", this.traceObjectId, ex);
+                            failedIteration.set(true);
+                        } else {
+                            failedIteration.set(false);
+                        }
+                        return null;
+                    });
+        }
+
+        private CompletableFuture<EventsReadAndTruncationPoints> readEvents() {
+            AtomicLong startOffset = new AtomicLong(0L);
+            AtomicInteger readLength = new AtomicInteger(0);
+            return CompletableFuture.supplyAsync(() -> {
+                                        // Need to get the start and end offsets to properly create async read request.
+                                        startOffset.set(segment.getInfo().getStartOffset());
+                                        readLength.set((int) Math.min(segment.getInfo().getLength() - startOffset.get(), MAX_TOTAL_EVENT_SIZE));
+                                        return segment.read(startOffset.get(), readLength.get(), containerOperationTimeout);
```

Comment:
I am not sure how we are dealing with events that fall across the boundary of two such reads.

It is possible that one read only reads the event bytes partially and then we need to issue next read and stitch together the remaining part and then only we can deserialize that event..

Reply:
Yes, the `deserializeEvents()` method is intended to handle any serialization error (`IOException`) or partial read (`BufferView.Reader.OutOfBoundsException`), but only valid events will be processed and the truncation point is based only on valid events.